python/ray/experimental/channel/serialization_context.py

```diff
@@ -9,11 +9,30 @@ class _SerializationContext:
     def __init__(self):
         self.torch_device: Optional["torch.device"] = None
         self.use_external_transport: bool = False
+        self.tensors: List["torch.Tensor"] = []
+
+    def set_use_external_transport(self, use_external_transport: bool) -> None:
+        self.use_external_transport = use_external_transport
 
     def set_torch_device(self, torch_device: "torch.device") -> None:
         self.torch_device = torch_device
 
-    def serialize_tensor(self, tensor: "torch.Tensor") -> "np.ndarray":
+    def reset_tensors(self, tensors: List["torch.Tensor"]) -> List["torch.Tensor"]:
+        prev_tensors = self.tensors
+        self.tensors = tensors
+        return prev_tensors
+
+    def serialize_tensor(self, tensor: "torch.Tensor") -> Union[int, "np.ndarray"]:
+        if self.use_external_transport:
+            # Add the actual tensor to a buffer. The buffer of tensors should
+            # later be popped by the caller and sent via external transport.
+            self.tensors.append(tensor)
```

Comment:
Does it mean that:
(1) if we use NCCL to transfer data, both GPUs (sender and receiver) can access `self.tensors`?
(2) if we use NCCL to transfer an integer, we don't need to serialize the integer?"

For the case `self.use_external_transport: False`, can we transfer an integer from `GPU -> CPU -> shared memory`, and then have the receiver GPU deserialize it by accessing `self.tensors[integer]` instead of transferring a numpy array?

Reply:
I'm not sure I understand you. We never send the integer placeholder over NCCL, only tensors. The protocol here is:

(1) Extract list of tensors. Replace with an integer as a placeholder. Each tensor's placeholder is its index into the list of tensors.
(2) On the sending GPU, use NCCL to send the list of tensors, in order.
(3) On the receiving GPU, use NCCL to receive the list of tensors, in order.
(4) Deserialize the rest of the data. If we see an integer placeholder, replace it with the corresponding received tensor.

Reply:
I think I understand how serialization/deserialization works after chatting with @rynewang. I will write it down and check with you tomorrow to see if my understanding is correct.